### PR TITLE
Remove MainFrame, merge it into GameRunner

### PR DESCRIPTION
**Pull request by DanVanAtta**
While tugging on various threads, I saw that we had a number of models in MainFrame. We also made it a singleton and added main frame open/close methods to it. I begun this by trying to mvoe the models out of MainFrame. I saw that the caller of MainFrame, GameRunner, could just as easily create and own the model objects. I started down this road, and MainFrame was eventually left with nothing except very boiler plate JFrame code. At this point I realized that GameRunner could keep track of MainFrame as a regular frame. For now the new variables in GameRunner are mostly static, but there is a path to avoid this.

Also of note, GameRunner is now stateful. The main(String[]) method just calls a new GameRunner.

### CODE REVIEW

**Comment by RoiEXLab** [comment]
I'd rather like to have the UI Code split from the main method.
Not sure if this goes beyond the scope of this PR

File: src/main/java/games/strategy/engine/data/properties/FileProperty.java [hunk 1]
```diff
@@ -167,7 +166,7 @@ public String getDescription() {
         return Arrays.toString(acceptableSuffixes);
       }
     });
-    final int rVal = fileChooser.showOpenDialog(MainFrame.getInstance());
+    final int rVal = fileChooser.showOpenDialog(null);
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/data/properties/FileProperty.java]: Not a big deal, but I just wanted to point out that using a `null` parent frame will center child windows to the screen rather than the TripleA application.  This only becomes an issue when you're running multiple (non-maximized) instances of TripleA.

I typically do this when testing network compatibility.  I'll open two TripleA applications and run them side-by-side on my single-monitor system.  It gets confusing when one instance pops up a window and it's centered relative to the screen rather than the originating instance.  Then I have to spend a few seconds figuring out which application it is associated with (usually clear from the title bar).

With that said, I'm wondering if we should be striving to use a non-`null` frame window wherever possible instead of going in the other direction?  Should the `GameRunner#mainFrame` field be exposed through a public static method, e.g. `GameRunner#getMainFrame()`?  I suspect there will be other changes in this PR whereby the main frame reference is replaced with `null`.

File: src/main/java/games/strategy/engine/framework/GameRunner.java [hunk 2]
```diff
@@ -110,21 +125,31 @@
...
       return;
     }
+    SwingUtilities.invokeLater(LookAndFeel::setupLookAndFeel);
+    ErrorConsole.getConsole();
     // do after we handle command line args
     Memory.checkForMemoryXMX();
 
-    SwingUtilities.invokeLater(LookAndFeel::setupLookAndFeel);
+    gameSelectorModel = new GameSelectorModel();
+    setupPanelModel = new SetupPanelModel(gameSelectorModel);
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/framework/GameRunner.java]: Both `gameSelectorModel` and `setupPanelModel` are accessed by multiple threads.  Instead of initializing both `gameSelectorModel` and `setupPanelModel` here and worrying about making them `volatile`, it looks like they can be made `final` and initialized via field initializers or in the class initializer.  For example:

```java
private static final GameSelectorModel gameSelectorModel = new GameSelectorModel();
private static final SetupPanelModel setupPanelModel = new SetupPanelModel(gameSelectorModel);
```

File: src/main/java/games/strategy/engine/framework/GameRunner.java [hunk 3]
```diff
@@ -656,4 +702,41 @@ public static void exitGameIfFinished() {
...
+    }
+    // having an oddball issue with the zip stream being closed while parsing to load default game. might be caused by
+    // closing of stream while unloading map resources.
+    ThreadUtil.sleep(100);
+    setupPanelModel.showSelectType();
+    showMainFrame();
+  }
+
+  public static void quitGame() {
+    mainFrame.dispatchEvent(new WindowEvent(mainFrame, WindowEvent.WINDOW_CLOSING));
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/framework/GameRunner.java]: Might want to guard against this running off the EDT as you did in `clientLeftGame()`.  Currently, the only caller (in `MainPanel`) is already running on the EDT, but you never know what someone will do in the future. :smile:

**Comment by DanVanAtta** [comment]
@RoiEXLab  
> I'd rather like to have the UI Code split from the main method.
Not sure if this goes beyond the scope of this PR

 GameRunner previously invoked `MainWindow` which then loaded up a bunch of model objects. Here we move those model objects out of `MainWindow` to `GameRunner`. `MainWindow` then was converted to a plain old `JFrame` object (what remained of `MainWindow` when removing the model is now in `SwingComponents`)

So I agree with the sentiment @RoiEXLab , the goal here is to move model code out of UI, and when we do that for MainWindow we see there is basically no UI code there (so MainWindow.java goes away)

File: src/main/java/games/strategy/engine/data/properties/FileProperty.java (hunk 1 above)
Comment (DanVanAtta) [inline src/main/java/games/strategy/engine/data/properties/FileProperty.java]: For multiple screens, it could get quite annoying, particularly if not both screens are visible at the same time. I pushed a commit to this branch that will restore the main frame reference: 0c6a9f1

Instead of a direct reference to `mainFrame` being leaked everywhere, instead `GameRunner`, which owns the `mainFrame` reference will be an intemediary for these swing calls. This way `GameRunner` can hide the `mainFrame` object reference from the rest of the system. Still pretty coupled, but a little bit less so and I hope a good step for further refactoring iterations.

Comment (ssoloff) [inline src/main/java/games/strategy/engine/data/properties/FileProperty.java]: Much better than my suggestion to simply expose `GameRunner#mainFrame`.  I like the direction this is going.  Thanks for the extra effort to do this, @DanVanAtta!

File: src/main/java/games/strategy/engine/framework/GameRunner.java [hunk 4]
```diff
@@ -150,6 +179,105 @@ public static void main(final String[] args) {
     }
   }
 
+  public static FileDialog createFileDialog( ) {
+    return new FileDialog(mainFrame);
+  }
+
+  public static FileDialog newFileDialog() {
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/framework/GameRunner.java]: Is there a semantic difference between `createFileDialog()` and `newFileDialog()`?

File: src/main/java/games/strategy/engine/framework/GameRunner.java (hunk 3 above)
Comment (DanVanAtta) [inline src/main/java/games/strategy/engine/framework/GameRunner.java]: I'm not sure about adding something that is not necessary. It is debatable to add it here. Should we always opportunistically wrap any such code in an EDT guard? My default reaction is to say no and go with the (simpler) fewer lines of code.If this were something where there were a silent error, then I would be inclined otherwise (not the case here since we would get an error if we invoke this on the wrong thread).  So, I am on the fence on this one, I would rather not see us pepper almost any such code with EDT guards (but maybe that would be a good thing... debatable)

Comment (ssoloff) [inline src/main/java/games/strategy/engine/framework/GameRunner.java]: Yeah, it's not as clear-cut to me anymore due to all the new EDT-bound methods that were added in the most recent commits (e.g. `createFileDialog()`, `showFileChooser()`, etc.).

I guess what was niggling me before was, as a dev, how do I know that `quitGame()` is required to run on the EDT?  The method name doesn't suggest anything related to the UI, and `GameRunner` is not in a UI-specific package, so I probably would have assumed I could just call it from anywhere.  As you said, I will find out during testing that I called it from the wrong thread and adjust appropriately. :smile:

However, I would argue that it is clear that the new methods you just added **are** UI specific since most of them have UI types in their signatures or their names are suggestive of UI behavior.  Therefore, I would think twice before calling them from a non-UI thread and would probably avoid threading issues from the beginning.

**Bottom line is you've convinced me that the code change to `quitGame()` is not necessary.**  Maybe a Javadoc stating that the method needs to be run from the EDT would be helpful, but I'm not suggesting it be done because I realize the recent changes to `GameRunner` are still a work in progress.  Based on your [earlier comment](https://github.com/triplea-game/triplea/pull/1855#issuecomment-307584978), it looks like the UI code will most likely be moved out of `GameRunner` to a UI-specific class in the future anyway.  That's good enough for me. :smile:

File: src/main/java/games/strategy/engine/framework/GameRunner.java (hunk 4 above)
Comment (DanVanAtta) [inline src/main/java/games/strategy/engine/framework/GameRunner.java]: No, thanks - fixed in 2570de6

File: src/main/java/games/strategy/ui/SwingComponents.java [hunk 5]
```diff
@@ -101,8 +112,8 @@ public static ButtonGroup createButtonGroup(final JRadioButton... radioButtons)
     return group;
   }
 
-  public static JDialog newJDialog(String title) {
-    JDialog dialog = new JDialog(MainFrame.getInstance(), title);
+  public static JDialog newJDialog(final String title) {
+    final JDialog dialog = new JDialog((JFrame) null, title);
```
Comment (ssoloff) [inline src/main/java/games/strategy/ui/SwingComponents.java]: Note that we're still losing the main frame reference here.

This code is currently only used to display the Unit Help and Game Notes dialogs (available from the Help menu).  I tested what happens with the actual main frame reference (I temporarily made `GameRunner#mainFrame` public), and the dialogs are still not centered relative to the window unless one adds a call to `setLocationRelativeTo()`.  However, adding a call to `setLocationRelativeTo()` in `newJDialog()` may be too early, as it shouldn't usually be done until the dialog has been `pack`ed.  Upon further inspection, both locations that use `newJDialog()` actually have calls to `setLocationRelativeTo()` that are commented out (presumably because they didn't have access to the correct frame window at the time).

So, I'm not sure it's important to preserve the main frame reference for dialogs created by `newJDialog()`.  I'm willing to leave it up to you @DanVanAtta to decide if you want to extract a `GameRunner#newJDialog()` method or simply leave it as-is.  (Actually, as modeless is the default mode for `JDialog`, another possibility is to just inline the two existing calls as `new JDialog((JFrame) null, <title>)` and remove the `newJDialog()` method completely.)

**Comment by DanVanAtta** [comment]
Last few items were addressed, squashing and merging now.

**Comment by ssoloff** [comment]
Sorry, didn't get to this until now.  Thanks for making those extra changes, @DanVanAtta!
